nsis/src/nsis/file/NsisExecutable.java

```diff
@@ -98,56 +103,32 @@ private void initScriptHeader() throws IOException, InvalidFormatException {
 	 * Attempt to decompress the data from the reader. Supports LZMA algorithm. Will
 	 * eventually support Bzip2 and Zlib.
 	 * 
+	 * @param offset, the offset at which the compressed data can be found
 	 * @throws IOException
 	 */
-	private byte[] decompressData() throws IOException {
-		this.reader.setPointerIndex(this.headerOffset + NsisScriptHeader.getHeaderSize());
-		byte compressionByte = this.reader.readNextByte();
-		if (NsisConstants.COMPRESSION_LZMA == compressionByte) {
-			long compressedDataOffset;
-			int compressedDataLength;
-			byte[] compressedData;
-			int dictionarySize = this.reader.readNextInt();
-			compressedDataOffset = this.headerOffset + NsisScriptHeader.getHeaderSize()
-					+ NsisConstants.COMPRESSION_LZMA_HEADER_LENGTH;
-			// Flip the MSB to get the length
-			compressedDataLength = (this.scriptHeader.compressedHeaderSize & ~FLAG_IS_COMPRESSED)
-					- NsisConstants.COMPRESSION_LZMA_HEADER_LENGTH;
-			compressedData = this.reader.readByteArray(compressedDataOffset, compressedDataLength);
-			return decompressLZMA(compressedData, compressionByte, dictionarySize);
-		} else if (NsisConstants.COMPRESSION_BZIP2 == compressionByte) {
-			// TODO Bzip2 decompress
-			System.out.println("Decompress Bzip");
-			return new byte[0];
-		} else {// TODO find a was to identify Zlib compressed
-			// TODO Zlib decompress
-			System.out.println("Decompress Zlib");
-			return new byte[0];
+	private NsisDecompressionProvider decompressData(long offset) throws IOException {
+		InputStream compressedInputStream = this.reader.getByteProvider().getInputStream(offset);
+		if ((this.scriptHeader.compressedHeaderSize & FLAG_IS_COMPRESSED) != 0) { // Check if MSB is
+			// set
+			this.reader.setPointerIndex(offset);
+			byte compressionByte = this.reader.readNextByte();
```

Comment:
I don't think a peak is necessary here, because I already have the compressedInputStream (line 110) ready and that object includes the propByte. So even if other algorithms include that byte in the data that has to be decompressed, I can just pass the compressedInputStream. That is also why I do a compressedInputStream.skip() at line 117. What do you think about that?

Reply:
I don't think you need to call skip at all. Using a `WrappedByteProvider` should remove the need for the skip.

Algorithm:
- peek byte
- byte is lzma
  - read propbyte
  - read dictionary
  - get compressed data size
  - create `WrappedByteProvider` that is the compressed bytes
  - create `LzmaDecompressionProvider(byteProvider, propByte, dictionarySize)`